### DISCUSSION TITLE
callback for participant node

### DISCUSSION THREAD

**Discussion by txlei** [osrf/rmf_core#288] [Q&A] [answered]
question about spinning nodes in different threads. 

context from https://github.com/osrf/rmf_core/issues/273 The number of obstacle participants are unknown and these would only be added dynamically when a custom msg is received in a subscriber topic, i.e. `ObstacleNode::make(nodeName, position, duration)` function called in the topic_callback function of the subscriber. 

since the `make` function` has a callback function defined from using `async_make_participant`, I need to have the obstacleNode spinning to wait for the participant to be ready while the subscriber is running its callback. side note: i would also be implementing some update and delete function for each participant so the nodes have to be assessed later on. 

Scour the answers.ros page and tried these but to no avail:
- implemented a `rclcpp::executors::SingleThreadedExecutor` in the `ObstacleNode::make` method.  Using `add_node(my_obstacle_node)` and `spin`
- the ros2 examples on `MultiThreadedExecutor` showed examples where the nodes are known in the first place. 
Though the new nodes created are still alive, the callback function is not running. 

concurrently trying to understand the concepts behind threading too so any advice would be very helpful!

I notice that the design seem to have many nodes created, eg. adapters and participants, which means each one have to be a single process. wondering if there are any limitations to this.

**Comment by mxgrey** [comment, marked as answer]
I'm not 100% sure if I'm interpreting your description correctly. Whenever possible, linking to the code in question is always appreciated to cut away any ambiguity in interpretation, but I'll try my best based on the impressions I'm getting.

It sounds like you're creating a whole new node for each schedule participant. If so, that's not necessary at all and might cause a lot of extraneous DDS traffic for performing discovery.

You can create a single `rmf_traffic_ros2::schedule::Writer` that's attached to a single node, and you can make as many schedule participants with that one writer as you'd like. These can be generated dynamically at any time, even in the middle of runtime.

You don't actually have to worry too much about threading because that's mostly taken care of for you internally by `rmf_traffic_ros2`. The only threading detail I would advise caution on is that the callback that you give to `async_make_participant` may be executed simultaneously in a different thread from the thread that your ROS2 callbacks are being run in, even if you use a single threaded executor. Therefore, when you receive the participant in the callback, you should use a mutex to protect whatever container you store that participant in. Maybe you'll store it in a vector? Or a map? Whatever you use, that container needs to have a mutex that gets locked before each time anyone accesses or modifies the container.

But you shouldn't need to do anything with executors. A single threaded executor should suit your use case just fine.

**Reply by txlei** [reply]
thanks for the clarification. previously I thought that each participant require a node. Able to create multiple participant now.

1. As for the participants declaration, is there other ways to "persist" it in the single node (with subscriber and writer) without declaring in the node class. Current implementation is not ideal as I need to specify the number of participant I am expecting. 

SuTrafficNode.cpp
```cpp
class SuTrafficNode : public rclcpp::Node
{
public:
    SuTrafficNode(const rclcpp::NodeOptions& options);
    rmf_traffic_ros2::schedule::WriterPtr writer;
    rmf_utils::optional<rmf_traffic::schedule::Participant> participant[100]; //<------- here
```

main.cpp
```cpp
void create_participant_custom_method(...){
   ...
    node->writer->async_make_participant(
        std::move(description),
        [id, t = std::move(t), map_name](rmf_traffic::schedule::Participant participant)
        {
            node->p[id] = std::move(participant);
            //participant.set(...) <----- participant unregistered immediately
            ...
```

2. could you explain the `Writer::Implementation::Transport` class? thinking of using the `unregister_participant` method. The goal is to unregister the participant after the trajectory duration has ended. 

3. Is it possible to update the inserted trajectory's Vector3d position during the start and finish time? Do I need to `rmf_traffic::Trajectory::erase(waypoint)` and `insert(...)` again?

Edit: I am using clear() trajectory and then creating a new participant again. But this is not v good approach I think.
